python_modules/dagster-pipes/dagster_pipes/__init__.py

```diff
@@ -971,15 +995,25 @@ def __enter__(self) -> "PipesContext":
         return self
 
     def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
-        self.close()
+        # expected to all be present or absent together
+        # https://docs.python.org/3/reference/datamodel.html#object.__exit__
+        if exc_type and exc_value and traceback:
+            exc = _pipes_exc_from_tb(TracebackException(exc_type, exc_value, traceback))
+        else:
+            exc = None
+        self.close(exc)
 
-    def close(self) -> None:
+    def close(
+        self,
+        exc: Optional[PipesException] = None,
+    ) -> None:
         """Close the pipes connection. This will flush all buffered messages to the orchestration
         process and cause any further attempt to write a message to raise an error. This method is
         idempotent-- subsequent calls after the first have no effect.
         """
         if not self._closed:
-            self._message_channel.write_message(_make_message("closed", {}))
+            payload = {"exception": exc} if exc else {}
+            self._message_channel.write_message(_make_message("closed", payload))
```

Comment:
As a follow up could we make enums for constants for these. These bare strings make me nervous.